README.md
# papertrail_remote_syslog Chef Cookbook

[![Build Status](https://travis-ci.org/rjocoleman/papertrail_remote_syslog.svg?branch=master)](https://travis-ci.org/rjocoleman/papertrail_remote_syslog)

A Chef cookbook to install [remote_syslog](https://github.com/papertrail/remote_syslog/) - a ruby daemon to transmit logs to papertrailapp.com

Currently supports only init and upstart.


# Attributes

Attributes listed with their default values:

```ruby
default['remote_syslog']['init_style'] = 'init'
default['remote_syslog']['conf']['files'] = []
default['remote_syslog']['destination']['host'] = 'logs.papertrailapp.com'
default['remote_syslog']['destination']['port'] = '12345'
default['remote_syslog']['hostname'] = node['hostname']
default['remote_syslog']['exclude_files'] = []
default['remote_syslog']['parse_fields'] = nil
default['remote_syslog']['prepend'] = nil
default['remote_syslog']['exclude_patterns'] = []
```

`node['remote_syslog']['conf']['files']`, `node['remote_syslog']['conf']['exclude_files']` and `node['remote_syslog']['conf']['exclude_patterns']` are all arrays.
These attributes are transformed into the [yaml config file](https://github.com/papertrail/remote_syslog/blob/master/examples/log_files.yml.example.advanced).


# Recipes

* `papertrail_remote_syslog::default` meta recipe to install, configures and enables+runs upstart/init scripts.
* `papertrail_remote_syslog::configure` just create configuration file from attributes.
* `papertrail_remote_syslog::service` just enabled+run upstart/init script.


## Contributing

* File bug reports via GitHub issues.
* Pull requests are welcome.

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Attribution

This cookbook was based on and forked from the [remote_syslog cookbook](https://github.com/bbg-cookbooks/remote_syslog). It has since diverged and is not compatible.


CHANGELOG.md
## 0.3.3

* Fix incorrect config file generation
* Add test to ensure configure file is valid (to sample known good)
* Style improvements (Rubocop)


## 0.3.2

* update to newest remote_syslog example init script
* add new remote_syslog config setting - `new_file_check_interval`
* convert upstart to a template, add support for multiple executable dirs


## 0.3.1

* Rubocop style cleanup
* Simplify config yaml file generation
* Remove unused attributes


## 0.3.0

* Forked from upstream.
* Temporarily removed Jamie tests pending re-implementation with kitchen ci.
* Added support for upstart init scripts.
* Created README.md
* Refactored attributes to be consistent with newer versions of Chef.
* Changed the meta recipe.
* Re-do Chefspec from scratch (Chefspec 3.0).
* Add Travis CI


#### Forked from upstream.

Prior to v0.3.0 this was https://github.com/bbg-cookbooks/remote_syslog

## 0.2.0:

* Implementation of all current published remote_syslog configuration options

## 0.1.0:

* Initial release of remote_syslog


spec/unit/recipes/default_spec.rb
require_relative '../../spec_helper'

describe 'papertrail_remote_syslog::default' do
  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }

  it 'installs remote_syslog gem' do
    expect(chef_run).to install_gem_package('remote_syslog')
  end
end


spec/unit/recipes/service_spec.rb
require_relative '../../spec_helper'

describe 'papertrail_remote_syslog::service' do
  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }

  it 'enable remote_syslog service' do
    expect(chef_run).to enable_service('remote_syslog')
  end

  it 'start remote_syslog service' do
    expect(chef_run).to start_service('remote_syslog')
  end

  it 'creates init file' do
    expect(chef_run).to create_cookbook_file('/etc/init.d/remote_syslog')
  end

  it 'creates upstart file based on attrbiute override' do
    chef_run.node.set['remote_syslog']['init_style'] = 'upstart'
    chef_run.converge(described_recipe)

    expect(chef_run).to create_template('/etc/init/remote_syslog.conf')
  end

end


spec/unit/recipes/configure_spec.rb
require_relative '../../spec_helper'

describe 'papertrail_remote_syslog::configure' do
  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }
  let(:yml) { IO.read 'spec/data/valid.yml' }

  it 'creates log_files.yml config file' do
    expect(chef_run).to create_file('/etc/log_files.yml')
  end

  it 'creates a well formed config file' do
    chef_run.node.set['remote_syslog']['conf']['files'] = [
      '/var/log/httpd/access_log', '/var/log/httpd/error_log',
      '/opt/misc/*.log', '/home/**/*.log', '/var/log/mysqld.log',
      '/var/run/mysqld/mysqld-slow.log'
    ]
    # rubocop:disable LineLength
    chef_run.node.set['remote_syslog']['destination']['host'] = 'logs.papertrailapp.com'
    # rubocop:enable LineLength
    chef_run.node.set['remote_syslog']['destination']['port'] = 12345
    chef_run.node.set['remote_syslog']['hostname'] = 'www42'
    chef_run.node.set['remote_syslog']['exclude_files'] = ['old', '200d']
    chef_run.node.set['remote_syslog']['parse_fields'] = 'syslog'
    chef_run.node.set['remote_syslog']['prepend'] = '0xDEADBEEF: '
    chef_run.node.set['remote_syslog']['exclude_patterns'] = [
      'exclude this', 'd+ things'
    ]
    chef_run.node.set['remote_syslog']['new_file_check_interval'] = 5
    chef_run.converge(described_recipe)

    expect(chef_run).to render_file('/etc/log_files.yml').with_content(yml)
  end
end


recipes/configure.rb
# # see README - demonstrates all optional arguments and more glob formats
# files:
#   - /var/log/httpd/access_log
#   - /var/log/httpd/error_log
#   - /opt/misc/*.log
#   - /home/**/*.log
#   - /var/log/mysqld.log
#   - /var/run/mysqld/mysqld-slow.log
# exclude_files:
#   - old
#   - 200\d
# hostname: www42  # override OS hostname
# parse_fields: syslog   # predefined regex name or double-quoted regex
# prepend: '0xDEADBEEF: '  # prepend this before every log message
# exclude_patterns:
#   - exclude this
#   - \d+ things
# destination:
#   host: logs.papertrailapp.com
#   port: 12345   # NOTE: change this to YOUR papertrail port!
# new_file_check_interval: 5 # Check every 5 seconds

config = {
  'files'                   => node['remote_syslog']['conf']['files'].to_a,
  'exclude_files'           => node['remote_syslog']['exclude_files'].to_a,
  'hostname'                => node['remote_syslog']['hostname'],
  'parse_fields'            => node['remote_syslog']['parse_fields'],
  'prepend'                 => node['remote_syslog']['prepend'],
  'exclude_patterns'        => node['remote_syslog']['exclude_patterns'].to_a,
  'destination'             => node['remote_syslog']['destination'].to_h,
  'new_file_check_interval' => node['remote_syslog']['new_file_check_interval']
}

file '/etc/log_files.yml' do
  owner 'root'
  group 'root'
  mode '0644'
  content config.to_yaml
end
